### PR TITLE
Tinlake V1

### PR DESCRIPTION
**Pull request by vedhavyas**

### CODE REVIEW

**Review by branan — APPROVED** [review]
As far as the merge itself goes, this looks clean to me.

We just need to coordinate this with Polkadot updates for the runtime.

File: pallets/loan/src/loan_type.rs [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+			LoanType::CreditLine(_) => None,
+			LoanType::CreditLineWithMaturity(clm) => Some(clm.maturity_date),
+		}
+	}
+
+	pub(crate) fn is_valid(&self, now: u64) -> bool {
+		match self {
+			LoanType::BulletLoan(bl) => bl.is_valid(now),
+			LoanType::CreditLine(cl) => cl.is_valid(),
+			LoanType::CreditLineWithMaturity(clm) => clm.is_valid(now),
```
Comment (mustermeiszer) [inline pallets/loan/src/loan_type.rs:46]: Maybe introduce a `LoanType` trait that boundles `is_valid`, `present_value` etc. together to an interface or are the arguments needed for the different loan-type to heterogenous for that?

File: primitives/tokens/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	() => {
+		pub struct TrancheToken<T>(core::marker::PhantomData<T>);
+
+		impl<T> pallet_tinlake_investor_pool::TrancheToken<T> for TrancheToken<T>
+		where
+			T: Config,
+			<T as Config>::PoolId: Into<u64>,
+			<T as Config>::TrancheId: Into<u8>,
+			<T as Config>::CurrencyId: From<CurrencyId>,
+		{
```
Comment (mustermeiszer) [inline primitives/tokens/src/lib.rs:27]: `From<CurrencyId>` refers to type of `CurrencyId` that we import in the context of this trait being used, right? I think we should have hidden imports here, that make it clear, to which `CurrencyId` type the `From` refers to!

File: runtime/common/src/currency.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	<T as Config>::TrancheId: Into<u8>,
+	<T as Config>::CurrencyId: From<CurrencyId>,
+{
+	fn tranche_token(
+		pool: <T as Config>::PoolId,
+		tranche: <T as Config>::TrancheId,
+	) -> <T as Config>::CurrencyId {
+		CurrencyId::Tranche(pool.into(), tranche.into()).into()
+	}
+}
```
Comment (mustermeiszer) [inline runtime/common/src/currency.rs:29]: Shouldn't we use our own marco here?

File: runtime/common/src/lib.rs [hunk 4]
```diff
@@ -117,6 +115,34 @@ pub mod types {
...
+		Copy,
+		Clone,
+		PartialEq,
+		Eq,
+		codec::CompactAs,
+		Debug,
+		TypeInfo,
+	)]
+	#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+	pub struct InstanceId(pub u128);
```
Comment (mustermeiszer) [inline runtime/common/src/lib.rs:145]: Do we wanna have a common `TrancheId` and `EpochId` to, if the rest is here?

**Review by mustermeiszer — APPROVED** [review]
Nice work, just minor remarks!

File: pallets/loan/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+			+ TypeInfo
+			+ IsType<ClassIdOf<Self>>;
+
+		/// The LoanId/InstanceId type
+		type LoanId: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ Copy
+			+ TypeInfo
+			+ From<u128>
```
[the remark this answers is not in the document]
Comment (vedhavyas) [reply pallets/loan/src/lib.rs:97]: yes. the we dont expect to offer 2^128 loans anyway 😉

File: pallets/loan/src/loan_type.rs (hunk 1 above)
Comment (vedhavyas) [reply pallets/loan/src/loan_type.rs:46]: no, make it a trait doen't help us with different func definition and args we need for speicifc loan types. 
That IMO would be too much abstraction

File: primitives/tokens/src/lib.rs (hunk 2 above)
Comment (vedhavyas) [reply primitives/tokens/src/lib.rs:27]: I'm not sure if we want to keep this here. I would prefer to bundle this as part of pool pallet. thoughts @branan ?

File: runtime/common/src/currency.rs (hunk 3 above)
Comment (vedhavyas) [reply runtime/common/src/currency.rs:29]: true! see the above. 
Also, macro doesn't seem of much use if the same impl is referred everywhere instead if generating one though macro
We can remove this macro TBH

File: pallets/loan/src/loan_type.rs (hunk 1 above)
Comment (mustermeiszer) [reply pallets/loan/src/loan_type.rs:46]: Yeah, reasonable.

File: runtime/common/src/lib.rs (hunk 4 above)
Comment (vedhavyas) [reply runtime/common/src/lib.rs:145]: yes for sure. Maybe a seperate ticket?

File: libs/common-traits/Cargo.toml [hunk 6]
```diff
@@ -21,6 +22,7 @@ sp-std = { git = "https://github.com/paritytech/substrate", default-features = f
 [features]
 default = ['std']
 std = [
+    "serde/std",
```
[the remark this answers is not in the document]
Comment (vedhavyas) [reply libs/common-traits/Cargo.toml:25]: I think so. I'm using this when its std, so I bundled serde in the same feature set.

**Pull request merged**